src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPoolWorkQueue.cs

```diff
@@ -723,14 +723,27 @@ internal static void TransferAllLocalWorkItemsToHighPriorityGlobalQueue()
             // Pop each work item off the local queue and push it onto the global. This is a
             // bounded loop as no other thread is allowed to push into this thread's queue.
             ThreadPoolWorkQueue queue = ThreadPool.s_workQueue;
+            bool anyWorkItemMoved = false;
             while (tl.workStealingQueue.LocalPop() is object workItem)
             {
-                queue.highPriorityWorkItems.Enqueue(workItem);
-            }
+                try
+                {
+                    queue.highPriorityWorkItems.Enqueue(workItem);
+                }
+                catch (Exception ex)
+                {
+                    // Typically would be an OOM, but since we have lost a work item, fail-fast on any exception
```

Comment:
Do we FailFast in other cases where a work item might be lost?

Reply:
Sort of, there are cases like [this](https://github.com/dotnet/runtime/blob/ea43e17c953a1230667c684a9f57d241e8a95171/src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPoolWorkQueue.cs#L1152) and [this](https://github.com/dotnet/runtime/blob/ea43e17c953a1230667c684a9f57d241e8a95171/src/libraries/System.Private.CoreLib/src/System/Threading/PortableThreadPool.IO.Windows.cs#L259) where a failure to add a work item would lead to an unhandled exception. In this case, the exception may transition user code and could get handled, so there's an explicit fail-fast here.

Reply:
Rather than fail fasting, should we instead push the popped work item back into the local queue we just dequeued from? That shouldn't require allocation as this is the only thread that could be adding to that queue, and there had to be room for this work item because we just got it from the queue.

It seems problematic to introduce fail fasts here that could just happen in low-memory situations.

Reply:
Yea I considered that and figured it would be uncommon enough. I don't feel strongly either way though, I'll switch to this alternative.

Reply:
The other thing I was thinking was in these uncommon cases a fail-fast might be easier to diagnose than the deadlock-like issues.